### PR TITLE
Feature/upgrade libjpeg

### PR DESCRIPTION
**Pull request by VirgileD**
As explained yesterday, I'd to have the libjpeg shipped with dlib upgraded to the version 8d. It will allow me to load a jpeg image from a memory buffer (with jpeg_mem_src). Most of the system that already includes a non broken libjpeg are at the version 8d.

So this pull request is made of 2 commits:
- one that upgrade the libjpeg included
- one that adds an overloaded jpeg_load fonction that takes a buffer and a size as parameter.

### CODE REVIEW

File: dlib/image_loader/jpeg_loader.h [hunk 1]
```diff
@@ -95,6 +97,17 @@ namespace dlib
     {
         jpeg_loader(file_name).get_image(image);
     }
+    template <
+        typename image_type
+        >
+    void load_jpeg (
+        image_type& image,
+        unsigned char* imgbuff,
+        long imgbuffsize
```
[the remark this answers is not in the document]
Comment (VirgileD) [reply dlib/image_loader/jpeg_loader.h]: Ok, done.

**Comment by davisking** [comment]
Nice :)

**Comment by davisking** [comment]
@VirgileD I just noticed that this new libjpeg is unable to open some jpeg files.  For instance, if you run this example `./image_ex ../faces/Tom_Cruise_avp_2014_4.jpg` you will get the error `exception thrown: jpeg_loader: error while loading image`.  Is there a way to get it to work?

**Comment by VirgileD** [comment]
Damn. Not home this week-end, I will check that on monday. But I didnt see any param in the libjeg that allows to force open a file.

**Comment by davisking** [comment]
Yes, it doesn't seem like this should happen.  It does though ¯\_(ツ)_/¯

**Comment by VirgileD** [comment]
Hello. On my particular toolchain (msys2 on windows) the error is related to this
```C
#ifndef TEMP_DIRECTORY		/* can override from jconfig.h or Makefile */
#define TEMP_DIRECTORY  "/usr/tmp/" /* recommended setting for Unix */
#endif
```
in jmemname.c. But i'm not sure you had the same problem if you're on linux where the /usr/tmp should be ok.
I modified the jpeg_loader.cpp to output the jpeg error message, like for example in my case:
```bash
> image_ex.exe ..\faces\Tom_Cruise_avp_2014_4.jpg
exception thrown: jpeg_loader: error while loading image: Failed to create temporary file /usr/tmp/JPG1a11844
```
Unfortunatly I deleted my previous branch so I made another PR: https://github.com/davisking/dlib/pull/1773 ; so you can tell me if you have the same error as mine.

Then, Imust admit that I'm not too sure how to deal with it: modify the libjpeg code, modify jconfig.h (how?) modify CmakeLists.txt (which one, how?). My best bet is to add:
```cmake
add_definitions(-DTEMP_DIRECTORY=$ENV{TMP})
```
in the CMakeLists.txt just after 
```cmake
           # If we can't find libjpeg then statically compile it in.
            add_definitions(-DDLIB_JPEG_STATIC)
```
but I'm pretty sure this will not be cross-platform.
What do you think?

**Comment by davisking** [comment]
It seems like you used the `jconfig.h` that was in dlib already?  That's probably why it's weird.  You need to configure whatever version of libjpeg you used and start with that jconfig.h as it probably introduces new settings.  In particular, I would be pretty surprised if the default configuration set it up to randomly write temp files to disk.  That is sort of crazy.

**Comment by davisking** [comment]
Actually, it's because there are three versions of that backing store thing in libjpeg, you are only supposed to link in one but you had it setup to link them all. Which ends up just using one of them, the wacky temp file one.  I pushed a change that makes libjpeg use `external/libjpeg/jmemnobs.c` which uses normal RAM, rather than disk for all this stuff.  I also removed a bunch of seemingly unneeded other things.

**Comment by VirgileD** [comment]
yes, I was almost there. Thank you, and sorry you finally had to work on that stuff (you must feel like, https://www.youtube.com/watch?v=eMvYJsjwT3Y ;-))

**Comment by davisking** [comment]
lol, no worries :)

**Pull request merged**